### PR TITLE
Modals service refactorization - https://particl.atlassian.net/browse/PD-152

### PR DESCRIPTION
**Pull request by cis-shubham-t**

### CODE REVIEW

**Comment by cis-shubham-t** [comment]
@kewde please have a look I have done some refactoring related to unlock modal only.

File: src/app/market/buy/checkout-process/checkout-process.component.ts [hunk 1]
```diff
@@ -227,7 +229,7 @@ export class CheckoutProcessComponent implements OnInit, OnDestroy {
   placeOrder() {
     if (this.rpcState.get('locked')) {
       // unlock wallet and send transaction
-      this.modals.open('unlock', { forceOpen: true, timeout: 30, callback: this.bidOrder.bind(this) });
+      this.modals.unlock({ forceOpen: true, timeout: 30, callback: this.bidOrder.bind(this) });
```
Comment (kewde) [inline src/app/market/buy/checkout-process/checkout-process.component.ts]:
```
this.modals.unlock({ forceOpen: true, timeout: 30}, (status) => {
  this.bidOrder()
});
```

Comment (cis-shubham-t) [reply src/app/market/buy/checkout-process/checkout-process.component.ts]: Sure, thanks

**Comment by HenkSwardt** [comment]
https://github.com/particl/particl-desktop/issues/1034

File: src/app/modals/modals-helper.service.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { Subject } from 'rxjs/Subject';
+import { Log } from 'ng2-logger';
+
+import { MatDialog, MatDialogRef } from '@angular/material';
+import { Observable } from 'rxjs/Observable';
+
+import { RpcStateService } from 'app/core/rpc/rpc-state/rpc-state.service';
+
+import { UnlockwalletComponent } from 'app/modals/unlockwallet/unlockwallet.component';
+import { ModalsConfig } from './models/modals.config.interface';
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: move ModalsConfig in here perhaps? Or are you going to make more models?

File: src/app/modals/modals-helper.service.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+@Injectable()
+export class ModalsHelperService implements OnDestroy {
+
+  // @TODO replace ModalsHelperService with ModalsService.
+
+  private log: any = Log.create('modals.service');
+  private destroyed: boolean = false;
+  private modelSettings: ModalsSettings = {
+    disableClose: true,
+    width: '100%',
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: rm

File: src/app/modals/modals-helper.service.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import { RpcStateService } from 'app/core/rpc/rpc-state/rpc-state.service';
+
+import { UnlockwalletComponent } from 'app/modals/unlockwallet/unlockwallet.component';
+import { ModalsConfig } from './models/modals.config.interface';
+
+interface ModalsSettings {
+    disableClose: boolean;
+    width: string;
+    height: string;
+    panelClass: string
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: remove width, height and panelClass

File: src/app/modals/modals-helper.service.ts (hunk 2 above)
Comment (kewde) [reply src/app/modals/modals-helper.service.ts]: ahh I see why - no prob.

Comment (cis-shubham-t) [reply src/app/modals/modals-helper.service.ts]: thanks :)

File: src/app/modals/models/modals.config.interface.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export interface ModalsConfig {
+  timeout?: number;
+  showStakeOnly?: boolean;
+  stakeOnly?: boolean;
+  autoClose?: boolean;
+  forceOpen?: boolean;
```
Comment (kewde) [inline src/app/modals/models/modals.config.interface.ts]: I think we can get rid of forceOpen?

Comment (kewde) [reply src/app/modals/models/modals.config.interface.ts]:
```
export interface UnlockModalConfig {
  timeout?: number;
  showStakeOnly?: boolean;
  stakeOnly?: boolean;
}
```

Comment (kewde) [reply src/app/modals/models/modals.config.interface.ts]: ^use a seperate config for the UnlockModal

File: src/app/modals/modals-helper.service.ts (hunk 3 above)
Comment (cis-shubham-t) [reply src/app/modals/modals-helper.service.ts]: why?

Comment (cis-shubham-t) [reply src/app/modals/modals-helper.service.ts]: another plan for the same?

**Comment by kewde** [comment]
@cis-shubham-t please fix tests.

**Comment by cis-shubham-t** [comment]
@kewde 
I fixed the test cases,
and I'm testing the current functionality as per the new changes.
So I can say the PR in the progress currently :)

File: src/app/modals/modals-helper.service.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      callback();
+    }
+  }
+
+    /**
+    * coldStack
+    * @param {string} type       type contains type of the modal.
+    */
+
+  coldStake(type: string) {
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: type not used?

File: src/app/modals/modals-helper.service.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    const dialogRef = this._dialog.open(ColdstakeComponent, this.modelSettings);
+    dialogRef.afterClosed().subscribe(() => {
+      this.log.d('coldStack modal closed');
+    });
+  }
+
+  syncing() {
+    if (!this.isOpen) {
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: use other fix we discussed if it works?

File: src/app/modals/models/modals.config.interface.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+export interface ModalsConfig {
+  timeout?: number;
+  showStakeOnly?: boolean;
```
Comment (kewde) [inline src/app/modals/models/modals.config.interface.ts]: Duplicate of UnlockModalConfig?

File: src/app/modals/models/unlock.modal.config.interface.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+export interface UnlockModalConfig {
+    timeout?: number;
+    showStakeOnly?: boolean;
```
Comment (kewde) [inline src/app/modals/models/unlock.modal.config.interface.ts]: Same as here.

File: src/app/modals/modals-helper.service.ts (hunk 6 above)
Comment (cis-shubham-t) [reply src/app/modals/modals-helper.service.ts]: currently not in use, But in future will use  `type`  as Cold? , ZapCold? or  RevertCold,
when using modal-helper service for all popup modals (Cold, ZapCold or RevertCold)

**Comment by AllienWorks** [comment]
Looks good to me & tested. Still needs code review.

File: src/app/modals/modals-helper.service.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    /* Hook BlockStatus -> update % `progress` */
+    this._blockStatusService.statusUpdates.asObservable().subscribe(status => {
+      this.progress.next(status.syncPercentage);
+    });
+
+    /* Hook wallet initialized -> open createwallet modal */
+    this.openInitialCreateWallet();
+
+    /* Hook daemon errors -> open daemon modal */
+    this._rpcState.errorsStateCall.asObservable()
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: remove this

File: src/app/modals/modals-helper.service.ts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+export class ModalsHelperService implements OnDestroy {
+
+  // @TODO replace ModalsHelperService with ModalsService.
+  private progress: Subject<Number> = new Subject<Number>();
+
+  /* True if user already has a wallet (imported seed or created wallet) */
+  public initializedWallet: boolean = false;
+  private log: any = Log.create('modals.service');
+  private destroyed: boolean = false;
+  public enableClose: boolean = true;
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: rm enableClose

File: src/app/modals/modals-helper.service.ts [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+  unlock(data: UnlockModalConfig, callback?: Function) {
+    if (this._rpcState.get('locked')) {
+      const dialogRef = this._dialog.open(UnlockwalletComponent, this.modelSettings);
+      if (!!data || callback) {
+        dialogRef.componentInstance.setData(data, callback);
+      }
+      dialogRef.afterClosed().subscribe(() => {
+        this.log.d('unlock modal closed');
+      });
+    } else if (!!callback) {
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: just do callback? I don't think there is a need for !! ?

File: src/app/modals/modals-helper.service.ts [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+  /**
+    * Unlock wallet
+    * @param {UnlockModalConfig} data       Optional - data to pass through to the modal.
+    */
+
+  unlock(data: UnlockModalConfig, callback?: Function) {
+    if (this._rpcState.get('locked')) {
+      const dialogRef = this._dialog.open(UnlockwalletComponent, this.modelSettings);
+      if (!!data || callback) {
```
Comment (kewde) [inline src/app/modals/modals-helper.service.ts]: same here !!data => data?

File: src/app/modals/bid-confirmation-modal/bid-confirmation-modal.component.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  @Output() onConfirm: EventEmitter<string> = new EventEmitter<string>();
+
+  public dialogContent: string;
+  public bidItem: Bid;
+  public country: string = '';
+
+  constructor(private dialogRef: MatDialogRef<BidConfirmationModalComponent>) { }
+
+  ngOnInit() {
+    // this.country = this.bidItem.ShippingAddress.country;
```
Comment (kewde) [inline src/app/modals/bid-confirmation-modal/bid-confirmation-modal.component.ts]: why is this // ?

File: src/app/modals/bid-confirmation-modal/bid-confirmation-modal.component.ts [hunk 15]
```diff
@@ -18,7 +18,7 @@ export class BidConfirmationModalComponent implements OnInit {
   constructor(private dialogRef: MatDialogRef<BidConfirmationModalComponent>) { }
 
   ngOnInit() {
-    // this.country = this.bidItem.ShippingAddress.country;
+    this.country = this.bidItem ? this.bidItem.ShippingAddress.country : '';
```
Comment (kewde) [inline src/app/modals/bid-confirmation-modal/bid-confirmation-modal.component.ts]: Does this work?

File: src/app/modals/bid-confirmation-modal/bid-confirmation-modal.component.ts (hunk 14 above)
Comment (kewde) [reply src/app/modals/bid-confirmation-modal/bid-confirmation-modal.component.ts]: kk

**Pull request merged**